### ISSUE TITLE
[Proposal] `.Context()` for `network.Conn` (or maybe just `swarm.Conn`)

### ISSUE THREAD

**Issue by MarcoPolo** [libp2p/go-libp2p#3116]
The proposal is for `conn.Context` to return a context that is cancelled when the connection is closed. User's can use the context to cleanup resources with `context.Afterfunc` or as a context for some other work.

## Motivation

It's important for users to clean up resources associated with a connection when that connection is closed. This is important for DoS resiliency and for performance. Right now the only way to do so is by registering a `network.Notifiee`, but that is relatively unergonomic.

There are network events, but those currently only tell you if you are (dis)connected to a peer, not if the connection you are referencing is closed.

Consider this common pattern seen in go-libp2p-kad-dht and boxo:

- Create a stream. This stream is implicitly tied to a connection.
- Store a reference to the stream to reuse the stream for multiple messages.
- subscribe to `event.EvtPeerConnectednessChanged` events to remove the
  stream reference when a peer disconnects.

This pattern is subtly incorrect. Imagine you had two connections to a peer, opened a stream on connection A, and then had the remote close that connection. You are still connected, but you haven't cleaned up your stream reference. This will prevent the GC from freeing the memory associated with that connection.

Another use case is to signal early cancellation of some work. For example, autonatv2 could use this to cancel it's dial back work if the connection is closed. Since, if the connection is closed, there's no way to report back the results of the dial back.

## Counter arguments

- Don't reuse streams and only keep well scoped references to a stream.
  - I'm for this actually. It may require changes to Yamux to support a configurable initial send buffer. More context here: <https://github.com/ipfs/boxo/issues/765>.
- It introduces another way to of cleaning up resources.
  - This is true. I'd argue this is better than using `network.Notifiee`. And, judging from existing usages, it's not often used.

## Prior Art

- In the Go standard library, `*http.Request` and `*tls.ClientHelloInfo` both have this method.
- In quic-go, the `quic.Connection` interface has this method.

## Details

As an implementation detail, the returned context will implement it's own `AfterFunc` method that does not spawn a goroutine. Users can still spawn their own goroutine if needed, but for the common case of dropping some state quickly it would be nice to avoid that overhead.

**Comment by burdiyan** [comment]
Just came here while looking for a way to have some metadata passed around the entire call path from the very top where the requests are handled. So, I was looking for either Conn-specific or (probably better) Stream-specific context, but unfortunately found none :(

**Comment by MarcoPolo** [comment]
Does wrapping the stream in your own new type work for your purpose?

**Comment by burdiyan** [comment]
Probably, but not entirely. Because this context won't be passed around all the libp2p and IPFS internals. For example I'd like to create a context on stream opening as some kind of a middleware, and add for example some ID in there, and then make make sure this context is passed to Bitswap and the underlying block service, such that I could read it in my custom block store implementation, and other related components. And that's currently not possible, because not all the components of libp2p/ipfs ecosystem are context-aware, and sometimes those that are simply using `context.Background()`.

It's a hard problem to solve — I get it :)

**Comment by MarcoPolo** [comment]
I don't think this proposal would solve your issue. This is not a way to attach data to a context, but rather attach some logic to the lifetime of a connection.

I'm not sure what the best solution for your problem is, but I wonder if something more explicit wouldn't be better.

**Comment by burdiyan** [comment]
I think it could have the potential to solve my issue, but it would require careful and accurate context passing in the overall ecosystem, and it's a pretty hard at this point I guess.

Basically for all the protocols, all the stream handlers that call some function requiring context would need to pass the upstream context from the connection, instead of creating a new one.

Here's an example of a Bitswap stream handler. It creates a separate context which is passed to the message receiver, which is then propagated everywhere context is ever taken. If libp2p conn and/or stream would have their own context, the bitswap handler would pass *it* instead of creating a new one. Exactly the same idea as with [http.Request.Context()](https://pkg.go.dev/net/http#Request.Context).

https://github.com/ipfs/boxo/blob/d16a4d5cfbc36d95b7bc8ba5b9a42c5d14d70285/bitswap/network/bsnet/ipfs_impl.go#L471
